### PR TITLE
Use dashmap+fxhash instead of flurry

### PR DESCRIPTION
**Pull request by alexkirsz**
While profiling, I noticed we spent a lot of time locked in `get_or_create_persistent_task` while flurry's HashMap was resizing itself[1]. I swapped it with dashmap and I'm seeing 8% improvements on our 10k modules benchmark.

image

image

### CODE REVIEW

File: crates/turbo-tasks-memory/src/memory_backend.rs [hunk 1]
```diff
@@ -100,7 +100,7 @@ impl MemoryBackend {
     }
 
     pub fn with_all_cached_tasks(&self, mut func: impl FnMut(TaskId)) {
-        for id in self.task_cache.pin().values() {
+        for id in self.task_cache.clone().into_read_only().values() {
```
Comment (jridgewell) [inline crates/turbo-tasks-memory/src/memory_backend.rs:103]: Do we need to clone if we're just immediately taking a read-only view?

**Review by jridgewell — APPROVED** [review]
What did you use to take the CPU profile?

**Comment by alexkirsz** [comment]
@jridgewell 
> What did you use to take the CPU profile?

I used https://github.com/cmyr/cargo-instruments. You'll need to build it yourself from the `update-deps` branch https://github.com/cmyr/cargo-instruments/pull/75 by running `cargo install --force --path .`. Then, you can instrument `next-dev` with:
```
cargo instruments -t time --bin next-dev --release -- ../triangles10k # or whatever app you want to test on
```

However, this only works if the program actually exits, and next-dev never does, so I usually add a `std::process::exit(0)` after one or two updates in `next-dev/src/main.rs`.

File: crates/turbo-tasks-memory/src/memory_backend.rs (hunk 1 above)
Comment (alexkirsz) [reply crates/turbo-tasks-memory/src/memory_backend.rs:103]: `into_read_only` consumes `self`, so we need to clone. However, `with_all_cached_tasks` is a debug helper so it doesn't really matter.

**Comment by alexkirsz** [comment]
It looks like the aforementioned improvements are also dependent on other changes I had stacked at the time I ran the benchmark, namely:
* Switch from the system allocator to [mimalloc](https://github.com/microsoft/mimalloc). We already had mimalloc in the dependency tree but we were not actually setting it up properly.
* Use FxHasher as DashMap's hasher. FWIW I tried this with flurry as well but it didn't improve anything.

**Review by jridgewell — APPROVED** [review]

File: crates/turbo-malloc/Cargo.toml [hunk 2]
```diff
@@ -9,8 +9,5 @@ autobenches = false
 [lib]
 bench = false
 
-[target.'cfg(not(target_os = "linux"))'.dependencies]
-mimalloc-rust = { version = "0.2" }
-
-[target.'cfg(all(target_os = "linux", not(target_arch = "aarch64")))'.dependencies]
-mimalloc-rust = { version = "0.2", features = ["local-dynamic-tls"] }
```
Comment (kwonoj) [inline crates/turbo-malloc/Cargo.toml:16]: This was intentional, as it broke next-swc target platforms. (Including above line 12)

**Review by ForsakenHarmony — APPROVED** [review]
LGTM if it doesn't make nft slower

**Comment by alexkirsz** [comment]
I've split the mimalloc change into https://github.com/vercel/turbo/pull/2467

**Review by jridgewell — APPROVED** [review]

**Pull request merged**